src/plugins/Wallet/web3.ts

```diff
@@ -1,60 +1,34 @@
 import Web3 from 'web3'
-import type { HttpProvider } from 'web3-core'
-import { PluginMessageCenter } from '../PluginMessages'
 import { sideEffect } from '../../utils/side-effects'
-import { currentEthereumNetworkSettings } from '../../settings/settings'
-import { getNetworkSettings } from './UI/Developer/EthereumNetworkSettings'
-import Services from '../../extension/service'
 import { OnlyRunInContext } from '@holoflows/kit/es'
 
 OnlyRunInContext('background', 'web3')
+import { WalletProviderType } from '../shared/findOutProvider'
+import { MetaMaskProvider } from '../../protocols/wallet-provider/metamask'
+import type { WalletProvider } from '../../protocols/wallet-provider'
+import { MaskbookProvider } from '../../protocols/wallet-provider/maskbook'
+import { unreachable } from '../../utils/utils'
+import { PluginMessageCenter } from '../PluginMessages'
+import { lastActivatedWalletProvider } from '../../settings/settings'
 
 export const web3 = new Web3()
-export const pool = new Map<string, HttpProvider>()
-
-let provider: HttpProvider
-
-export const resetProvider = () => {
-    const url = getNetworkSettings().middlewareAddress
-    provider = pool.has(url)
-        ? pool.get(url)!
-        : // more: https://github.com/ethereum/web3.js/blob/1.x/packages/web3-providers-ws/README.md
-          new Web3.providers.HttpProvider(url, {
-              timeout: 5000, // ms
-              // @ts-ignore
-              clientConfig: {
-                  keepalive: true,
-                  keepaliveInterval: 1, // ms
-              },
-              reconnect: {
-                  auto: true,
-                  delay: 5000, // ms
-                  maxAttempts: Number.MAX_SAFE_INTEGER,
-                  onTimeout: true,
-              },
-          })
-    if (pool.has(url)) provider.disconnect()
-    else pool.set(url, provider)
-    web3.setProvider(provider)
+let currentProvider: WalletProvider
```

Comment:
You can delete this local variable read it from `lastActivatedWalletProvider.value`.

Reply:
`lastActivatedWalletProvider: InternalSettings<WalletProviderType>` is an enum, `currentProvider` is an instance, they are different. The prior one is used on the initial stage to preserve the choice of last running